core/src/main/java/hudson/util/ProcessTree.java

```diff
@@ -167,6 +169,22 @@ public void killAll(Process proc, Map<String, String> modelEnvVars) throws Inter
     }
 
     /**
+     * @param p The process to check
+     * @return The first non-null {@link VetoCause} provided by a process killing veto extension. null if no one
+     *         objects.
+     */
+    protected @CheckForNull VetoCause getVeto(OSProcess p) {
+        for (ProcessKillingVeto vetoExtension : ProcessKillingVeto.all()) {
+            VetoCause cause = vetoExtension.vetoProcessKilling(p);
+            if (cause != null) {
+                LOGGER.finest("Killing of pid " + p.getPid() + " vetoed by " + cause.getSource().getClass().getName() + ": " + cause.getCause());
```

Comment:
IMO <code>VetoCause</code> should have a <code>getMessage()</code> (or something like that) method describing the veto reason.